store/store.go

```diff
@@ -440,24 +443,38 @@ func (bs *BlockStore) SaveBlock(block *types.Block, blockParts *types.PartSet, s
 		bs.saveBlockPart(height, i, part)
 	}
 
-	// Save block meta
-	blockMeta := types.NewBlockMeta(block, blockParts)
-	pbm := blockMeta.ToProto()
-	if pbm == nil {
-		panic("nil blockmeta")
-	}
-	metaBytes := mustEncode(pbm)
-	if err := bs.db.Set(blockMetaKey(height), metaBytes); err != nil {
-		panic(err)
-	}
-	if err := bs.db.Set(blockHashKey(hash), []byte(fmt.Sprintf("%d", height))); err != nil {
-		panic(err)
-	}
+	// Save the block meta data
+	tr.Append(func() {
+		blockMeta := types.NewBlockMeta(block, blockParts)
+		pbm := blockMeta.ToProto()
+		if pbm == nil {
+			panic("nil blockmeta")
+		}
 
-	// Save block commit (duplicate and separate from the Block)
-	pbc := block.LastCommit.ToProto()
-	blockCommitBytes := mustEncode(pbc)
-	if err := bs.db.Set(blockCommitKey(height-1), blockCommitBytes); err != nil {
+		metaBytes := mustEncode(pbm)
+		if err := bs.db.Set(blockMetaKey(height), metaBytes); err != nil {
+			panic(err)
+		}
+	}, blockMetaKey(height))
+
+	// Save block hashes
+	tr.Append(func() {
+		if err := bs.db.Set(blockHashKey(hash), []byte(fmt.Sprintf("%d", height))); err != nil {
+			panic(err)
+		}
+	}, blockHashKey(hash))
+
+	// Save block commits duplicate and separate from the Block)
+	tr.Append(func() {
+		pbc := block.LastCommit.ToProto()
+		blockCommitBytes := mustEncode(pbc)
+		if err := bs.db.Set(blockCommitKey(height-1), blockCommitBytes); err != nil {
+			panic(err)
+		}
+	}, blockCommitKey(height-1))
+
+	// Transact all the loaded transactions
+	if err := tr.Transact(bs.db); err != nil {
```

Comment:
i commented on the general approach in the other pr. but this pr still has exactly same issue as before. all writes except the last one do not guarantee to be on disk if method successfully returns and then machine crashes